### PR TITLE
Feature/worklist type state

### PR DESCRIPTION
**Pull request by luketlancaster**
- [x] Add the Clubhouse Story ID: [ch22236]
- [ ] Add QA tasks

### CODE REVIEW

File: src/js/apps/patients/worklist/filters_app.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ export default App.extend({
     if (this.groups.length < 2) return;
 
     const groups = this._getGroups();
-    const selected = groups.get(this.getState('groupId'));
+    const selected = groups.get(this.getState('groupId')) || groups.at(0);
```
Comment (paulfalgout) [inline src/js/apps/patients/worklist/filters_app.js:33]: We should really be making sure `groupId` is correct before this point rather than picking `at(0)`.. what is this use case accounting for?  _although_ I can see a edge case for when a user is removed from a group.. but otherwise this seems like there's a bug somewhere.

File: src/js/apps/patients/worklist/worklist_app.js [hunk 2]
```diff
@@ -134,7 +133,6 @@ export default App.extend({
     this.showSortDroplist();
     this.showTypeToggle();
     this.showTableHeaders();
-    this.startFiltersApp();
```
Comment (paulfalgout) [inline src/js/apps/patients/worklist/worklist_app.js]: why was this change necessary?  Work was done to make sure the filters app was updated as data changed.. so that'd need to come out.. but ideally it could be in onBeforeStart

File: src/js/apps/patients/worklist/worklist_app.js [hunk 3]
```diff
@@ -148,17 +146,18 @@ export default App.extend({
...
+    this.startFiltersApp();
   },
   getComparator() {
     const sortId = this.getState().getSort();
     return this.sortOptions.get(sortId).get('comparator');
   },
   getSortOptions() {
     if (this.getState().isFlowType()) {
-      return _.union(sortDueOptions, sortUpdateOptions);
+      return sortUpdateOptions;
```
Comment (paulfalgout) [inline src/js/apps/patients/worklist/worklist_app.js:155]: :+1: oops

File: src/js/apps/patients/worklist/worklist_app.js (hunk 2 above)
Comment (paulfalgout) [reply src/js/apps/patients/worklist/worklist_app.js]: ah... this is because the `filters.type` is changing outside of the filters app.  Ok.. well.. bad news... but I think it makes sense..  we need to move the toggle into the filters app.  Then all of the filters state is controlled by the same app.

File: src/js/views/patients/worklist/worklist_views.js [hunk 4]
```diff
@@ -108,6 +111,28 @@ const ListView = CollectionView.extend({
   },
 });
 
+const TypeToggleView = View.extend({
+  template: hbs`
+    <button class="button-secondary {{#unless isFlowList}}button--blue{{/unless}} worklist-list__toggle-actions js-toggle-actions">{{far "file-alt"}}{{ @intl.patients.worklist.layoutTemplate.actionsButton }}</button>{{~ remove_whitespace ~}}
```
Comment (paulfalgout) [inline src/js/views/patients/worklist/worklist_views.js]: i18n need updating

**Review by paulfalgout — CHANGES_REQUESTED** [review]
So I also just realized we'll have to extract the tooltip somehow from the layout.  We can't pass `isFlowList` to the layout as it would have to be re-rendered when the state changes.

File: src/js/apps/patients/worklist/filters_app.js (hunk 1 above)
Comment (luketlancaster) [reply src/js/apps/patients/worklist/filters_app.js:33]: if `groupId` is `null`, `groups.get` doesn't return anything. So I guess make the id for "All Groups" `all-groups` again?

File: src/js/apps/patients/worklist/worklist_app.js (hunk 2 above)
Comment (luketlancaster) [reply src/js/apps/patients/worklist/worklist_app.js]: I changed this back because the first click on a filter worked, but any subsequent click didn't fire off an event

File: src/js/apps/patients/worklist/filters_app.js (hunk 1 above)
Comment (paulfalgout) [reply src/js/apps/patients/worklist/filters_app.js:33]: yeah :-/    just check the places where `groupId` is falsey checked.. or is id `0` hacky?  kind of..

File: src/js/apps/patients/worklist/worklist_app.js (hunk 2 above)
Comment (paulfalgout) [reply src/js/apps/patients/worklist/worklist_app.js]: I think I get it.. so.. we should remove stateEvents from the filters app.. since those'll never happen.. and let's put this back in `onBeforeStart` and add this to the `if(isRestarting`  I _think_ that'll work and handle state changes during loading

Comment (luketlancaster) [reply src/js/apps/patients/worklist/worklist_app.js]: nope, no dice. Same behavior as before: first interaction with a filter component changes does what you think it would: sets the state in the store, new requests are fired off, list updates (though the 'reset button' doesn't show up until a page refresh). Then every other interaction with a filter only updates the droplist, but the store isn't being updated and so requests aren't being fired/the list doesn't get updated

File: src/js/apps/patients/worklist/filters_app.js (hunk 1 above)
Comment (luketlancaster) [reply src/js/apps/patients/worklist/filters_app.js:33]: so `0` is kinda hacky...though the `groups` right here is private to this app, and so we know that "All Groups" is the first thing (unless for some reason we get a request in the future from business to make "All Groups" the last thing in the droplist?)

File: src/js/apps/patients/worklist/worklist_app.js (hunk 2 above)
Comment (paulfalgout) [reply src/js/apps/patients/worklist/worklist_app.js]: back to the drawing board I guess

Comment (paulfalgout) [reply src/js/apps/patients/worklist/worklist_app.js]: Ok.. I'm pretty sure everything works as it was in my commit with the exception of this line which needs to change:
```js
  this.listenTo(filtersApp.getState(), 'change', ({ attributes }) => {
    this.setState({ filters: _.clone(attributes) });
  });
```
The issue is that we're changing an object for the same object so even though the data on the object changed the object referenced did not so no change event was occurring..  Generally why we don't nest state data (and want flat data from endpoints) but I dunno.. seemed cleaner this time :-/

**Comment by luketlancaster** [comment]
alrighty, take another look @paulfalgout. Still need to do the tooltips, but I _think_ this should be closer?

File: test/integration/services/patient-quick-search.js [hunk 5]
```diff
@@ -17,7 +17,9 @@ context('Patient Quick Search', function() {
 
         return fx;
       })
-      .visit();
+      .visit()
+      .navigate('/worklist/owned-by')
```
Comment (paulfalgout) [inline test/integration/services/patient-quick-search.js]: why does this need to navigate or wait at all?

File: test/integration/patients/worklist.js [hunk 6]
```diff
@@ -925,6 +880,32 @@ context('worklist page', function() {
...
+      .get('.app-frame__content')
+      .find('.table-list__item')
+      .last()
+      .should('contain', 'Updated Least Recent');
+
+    cy
+      .get('.worklist-list__filter-sort')
+      .click()
+      .get('.picklist')
+      .should('not.contain', 'Due Date: Oldest - Newest')
```
Comment (paulfalgout) [inline test/integration/patients/worklist.js]: could just have one `should` with `not.contain`, 'Due Date`

File: src/js/views/patients/worklist/filters_views.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      isFlowList: this.getOption('isFlowList'),
+    };
+  },
+  triggers: {
+    'click .js-toggle-actions': 'click:toggleActions',
+    'click .js-toggle-flows': 'click:toggleFlows',
+  },
+  onClickToggleActions() {
+    this.triggerMethod('toggle:listType', 'actions');
+    this.$el.find('button').toggleClass('button--blue');
```
Comment (paulfalgout) [inline src/js/views/patients/worklist/filters_views.js]:
```
ui: {
  buttons: 'button',
},
...
this.ui.toggleClass('button--blue')
```

File: src/js/views/patients/worklist/filters_views.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+const ClinicianClearButton = View.extend({
+  template: hbs`<button class="button-secondary worklist-list__clear-filter">{{ @intl.patients.worklist.filtersViews.clearClinicianFilter }}</button>`,
+  triggers: {
+    'click': 'click',
+  },
+});
+
+const TypeToggleView = View.extend({
+  template: hbs`
+    <button class="button-secondary {{#unless isFlowList}}button--blue{{/unless}} worklist-list__toggle-actions js-toggle-actions">{{far "file-alt"}}{{ @intl.patients.worklist.typeToggle.actionsButton }}</button>{{~ remove_whitespace ~}}
```
Comment (paulfalgout) [inline src/js/views/patients/worklist/filters_views.js]: worklist.filtersViews.typeToggleView.actionsButton

**Review by paulfalgout — CHANGES_REQUESTED** [review]
cool.. so pretty much just the tooltip

File: test/integration/services/patient-quick-search.js (hunk 5 above)
Comment (luketlancaster) [reply test/integration/services/patient-quick-search.js]: it sure doesn't

**Review by paulfalgout — CHANGES_REQUESTED** [review]
looks like a hair of coverage missing

File: src/js/apps/patients/worklist/filters_app.js [hunk 9]
```diff
@@ -69,7 +84,7 @@ export default App.extend({
     const groups = this.groups.clone();
 
     groups.unshift({
-      id: 'all-groups',
+      id: null,
       name: 'All Groups',
```
Comment (LisaManresa) [inline src/js/apps/patients/worklist/filters_app.js]: Should this be i18n-ified?

**Review by LisaManresa — APPROVED** [review]
One small thing, but probably not a blocker since it's already in develop 😬

File: src/js/apps/patients/worklist/filters_app.js (hunk 9 above)
Comment (luketlancaster) [reply src/js/apps/patients/worklist/filters_app.js]: yeah you're right it should 100% be i18n'd
